File: app/server/src/shared/services/logger/logger.service.ts
import { Injectable } from '@nestjs/common';
import * as winston from 'winston';
import { LOGGER_COMPONENT, LOGGER_DEFAULT_USER, LOGGER_PROJECT } from './consts/logger.const';
import { EcsLogPayload } from './types/ecs-log-payload.type';
import { LogContext } from './types/log-context.type';
import { LogUser } from './types/log-user.type';

type WinstonLevel = 'info' | 'warn' | 'error' | 'debug' | 'verbose';

@Injectable()
export class LoggerService {
  private readonly _winston: winston.Logger;
  private readonly _project: string = LOGGER_PROJECT;
  private readonly _component: string = LOGGER_COMPONENT;
  private readonly _environment: string = process.env.NODE_ENV ?? 'development';
  private readonly _user: LogUser = LOGGER_DEFAULT_USER;

  public constructor() {
    this._winston = winston.createLogger({
      level: 'verbose',
      format: winston.format.combine(
        winston.format.timestamp(),
        winston.format.errors({ stack: true }),
        winston.format.json(),
      ),
      transports: [new winston.transports.Console()],
    });
  }

  public log(message: string, logContext: LogContext): void {
    this._write('info', message, logContext);
  }

  public info(message: string, logContext: LogContext): void {
    this._write('info', message, logContext);
  }

  public warn(message: string, logContext: LogContext): void {
    this._write('warn', message, logContext);
  }

  public error(message: string, logContext: LogContext): void {
    this._write('error', message, logContext);
  }

  public debug(message: string, logContext: LogContext): void {
    this._write('debug', message, logContext);
  }

  public verbose(message: string, logContext: LogContext): void {
    this._write('verbose', message, logContext);
  }

  private _write(level: WinstonLevel, message: string, logContext: LogContext): void {
    const payload: EcsLogPayload = {
      project: this._project,
      component: this._component,
      environment: this._environment,
      message,
      user: this._user,
      logContext,
    };
    this._winston.log(level, payload);
  }
}

File: app/parser/shared/logger_service.py
"""ECS-format JSON logger service for the parser.

Mirrors the server-side LoggerService (Winston JSON output) so logs from both
components share a single shape. Five log levels are exposed (`log`/`info`,
`error`, `warn`, `debug`, `verbose`); each call accepts a message and a
log_context literal ("app-monitoring" or "app-workflow"). The project,
component, environment, and user fields are predefined on the service.
"""

import json
import logging
import os
import sys
from datetime import datetime, timezone
from typing import Literal


LogContext = Literal["app-monitoring", "app-workflow"]

_PROJECT = "buster"
_COMPONENT = "parser"

_DEFAULT_USER = {
    "email": "system@buster.local",
    "name": "system",
    "role": "system",
}

# Maps Winston-style level names to Python logging levels so the parser and
# server share the same vocabulary.
_LEVEL_MAP = {
    "info": logging.INFO,
    "warn": logging.WARNING,
    "error": logging.ERROR,
    "debug": logging.DEBUG,
    "verbose": logging.DEBUG,
}

_PYTHON_LEVEL_TO_WINSTON_LABEL = {
    "INFO": "info",
    "WARNING": "warn",
    "ERROR": "error",
    "DEBUG": "debug",
}


class _EcsJsonFormatter(logging.Formatter):
    """Renders the LogRecord's payload (a dict on `record.msg`) as a JSON line."""

    def format(self, record: logging.LogRecord) -> str:
        payload = record.msg if isinstance(record.msg, dict) else {"message": record.getMessage()}
        timestamp = datetime.now(timezone.utc).isoformat()
        level_label = _PYTHON_LEVEL_TO_WINSTON_LABEL.get(record.levelname, record.levelname.lower())
        return json.dumps({"timestamp": timestamp, "level": level_label, **payload})


class LoggerService:
    """Singleton-style service exposing Winston-equivalent log methods in ECS format."""

    def __init__(self) -> None:
        self._project = _PROJECT
        self._component = _COMPONENT
        self._environment = os.environ.get("PYTHON_ENV", "development")
        self._user = _DEFAULT_USER
        self._logger = logging.getLogger("buster.parser")
        self._logger.setLevel(logging.DEBUG)
        self._logger.propagate = False
        if not self._logger.handlers:
            handler = logging.StreamHandler(sys.stdout)
            handler.setFormatter(_EcsJsonFormatter())
            self._logger.addHandler(handler)

    def log(self, message: str, log_context: LogContext) -> None:
        self._write("info", message, log_context)

    def info(self, message: str, log_context: LogContext) -> None:
        self._write("info", message, log_context)

    def warn(self, message: str, log_context: LogContext) -> None:
        self._write("warn", message, log_context)

    def error(self, message: str, log_context: LogContext) -> None:
        self._write("error", message, log_context)

    def debug(self, message: str, log_context: LogContext) -> None:
        self._write("debug", message, log_context)

    def verbose(self, message: str, log_context: LogContext) -> None:
        self._write("verbose", message, log_context)

    def _write(self, level: str, message: str, log_context: LogContext) -> None:
        payload = {
            "project": self._project,
            "component": self._component,
            "environment": self._environment,
            "message": message,
            "user": self._user,
            "logContext": log_context,
        }
        self._logger.log(_LEVEL_MAP[level], payload)


logger_service = LoggerService()
